File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b)->(a[0]!=b[0]?a[0]-b[0]:a[1]-b[1]));
        List<int[]> list = new ArrayList<>();
        for (int[] interval : intervals)
        {
            if (list.isEmpty())
            {
                list.add(interval);
                continue;
            }
            //if overlap, merge
            int[] last = list.get(list.size()-1);
            int maxStart = Math.max(interval[0], last[0]);
            int minEnd = Math.min(interval[1], last[1]);
            if (minEnd-maxStart>=0)
            {
                list.remove(list.size()-1);
                int minStart = Math.min(interval[0], last[0]);
                int maxEnd = Math.max(interval[1], last[1]);
                list.add(new int[]{minStart, maxEnd});
            }
            else
            {
                list.add(interval);
            }
        }
        int[][] result = new int[list.size()][2];
        int index = 0;
        for (int[] interval : list)
        {
            result[index++] = interval;
        }
        return result;
    }
}

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        last_start = intervals[0][0]
        last_end = intervals[0][1]
        result = []
        for interval in intervals:
            # if overlap, keep going
            # if no overlap, add last interval
            ovrlp = min(last_end, interval[1])-max(last_start,interval[0])>=0
            if ovrlp:
                last_start = min(last_start, interval[0])
                last_end = max(last_end, interval[1])
            else:
                result.append([last_start, last_end])
                last_start = interval[0]
                last_end = interval[1]
        # add last interval
        result.append([last_start, last_end])
        return result

        